rclcpp/include/rclcpp/timer.hpp

```diff
@@ -198,6 +199,15 @@ class GenericTimer : public TimerBase
   void
   execute_callback() override
   {
+    // TODO(ivanpauno): Delete this workaround.
+    // The issue is that the multithreaded executor might schedule the timer
+    // more than once in different threads.
+    // Making `execute_callback` mutually exclusive and rechecking if the timer is ready
+    // avoids the issue.
+    std::lock_guard<std::mutex> lock(execution_mutex_);
```

Comment:
I think, with this fix there will be high possibility that multiple threads get the same timer, be exclusive on this lock. This would be a problem that if the user timer callback takes time. in that case, other threads just stops here and cannot to continue to get any executables.

Reply:
mmm, two timers being executed in parallel is an uncommon event, so contention will not happen frequently.
But yeah, maybe this is not great.

I'm still confused about https://github.com/ros2/rclcpp/issues/1374, the test checking that a timer is not executed twice is sometimes failing and I don't understand well why.

Reply:
> mmm, two timers being executed in parallel is an uncommon event

agree,

> the test checking that a timer is not executed twice is sometimes failing and I don't understand well why.

time is not executed twice, i think that the jitter because of lock contention.

if there is a lot of threads, it is hard to get lock https://github.com/ros2/rclcpp/blob/5851eebdda5cc2ed9186a034df0ef75ccfe032d3/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L106 because everyone else is just trying to loop to get lock . (this also consumes cpu resource.) https://github.com/ros2/rclcpp/blob/5851eebdda5cc2ed9186a034df0ef75ccfe032d3/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L79 I was thinking that we might solve this problem if we can provide higher priority to get lock https://github.com/ros2/rclcpp/blob/5851eebdda5cc2ed9186a034df0ef75ccfe032d3/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L106 to call https://github.com/ros2/rclcpp/blob/5851eebdda5cc2ed9186a034df0ef75ccfe032d3/rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp#L109

Reply:
Ah, I think I understand: we're seeing two timer callbacks being executed close because one was late and lose a period, but not because it was scheduled twice, right?
I have to recheck how the test works, but my description of https://github.com/ros2/rclcpp/issues/1374 is definitely wrong.

Reply:
> we're seeing two timer callbacks being executed close because one was late and lose a period, but not because it was scheduled twice, right?

yes, i do think so.